src/push-manifest.ts

```diff
@@ -0,0 +1,213 @@
+ *           },
+ *           ...
+ *         },
+ *         "page.html": {
+ *           "/css/page.css": {
+ *             "type": "style",
+ *             "weight": 1
+ *           },
+ *           ...
+ *         }
+ *       }
+ *
+ * NOTE(fks) 04-05-2017: Only weight=1 is supported by browsers at the moment.
+ * When support is added, we can add automatic weighting and support multiple
+ * numbers.
+ */
+export type ResourceType = 'document' | 'script' | 'style' | 'image' | 'font';
+export interface PushManifestEntry {
+  type?: ResourceType;
+  weight: 1;
+}
+export interface PushManifestEntryCollection {
+  [dependencyAbsoluteUrl: string]: PushManifestEntry;
+}
+export interface PushManifest {
+  [requestAbsoluteUrl: string]: PushManifestEntryCollection;
+}
+
+/**
+ * A mapping of file extensions and their default resource type.
+ */
+const extensionToTypeMapping: {[extname: string]: ResourceType} = {
+  '.css': 'style',
+  '.gif': 'image',
+  '.html': 'document',
+  '.png': 'image',
+  '.jpg': 'image',
+  '.js': 'script',
+  '.json': 'script',
+  '.svg': 'image',
+  '.webp': 'image',
+  '.woff': 'font',
+  '.woff2': 'font'
+};
+
+/**
+ * Get the default resource type for a file based on its extension.
+ */
+function getResourceTypeFromUrl(url: string): ResourceType|undefined {
+  return extensionToTypeMapping[path.extname(url)] || undefined;
+}
+
+/**
+ * Get the resource type for an import, handling special import types and
+ * falling back to getResourceTypeFromUrl() if the resource type can't be
+ * detected directly from importFeature.
+ */
+function getResourceTypeFromImport(importFeature: Import): ResourceType|
+    undefined {
+  const importKinds = importFeature.kinds;
+  if (importKinds.has('css-import') || importKinds.has('html-style')) {
+    return 'style';
+  }
+  if (importKinds.has('html-import')) {
+    return 'document';
+  }
+  if (importKinds.has('html-script')) {
+    return 'script';
+  }
+  // @NOTE(fks) 04-07-2017: A js-import can actually import multiple types of
+  // resources, so we can't guarentee that it's a script and should instead rely
+  // on the default file-extension mapping.
+  return getResourceTypeFromUrl(importFeature.url);
+};
+
+/**
+ * Create a PushManifestEntry from an analyzer Import.
+ */
+function createPushEntryFromImport(importFeature: Import): PushManifestEntry {
+  return {
+    type: getResourceTypeFromImport(importFeature),
+    weight: 1,
+  };
+}
+
+/**
+ * Analyze the given URL and resolve with a collection of push manifest entries
+ * to be added to the overall push manifest.
+ */
+async function generatePushManifestEntryForUrl(
+    analyzer: Analyzer, url: string, ignoreUrls?: string[]):
+    Promise<PushManifestEntryCollection> {
+      const analyzedDocument = await analyzer.analyze(url);
+      const analyzedImports = analyzedDocument.getByKind(
+          'import', {externalPackages: true, imported: true});
+      const analyzedImportUrls: PushManifestEntryCollection = {};
```

Comment:
Since we're constructing a JSON object I'd recommend keeping this as a JS object. Converting map-to-JS-object after the fact requires an extra step/iteration over the map and  the practical value added by a Map in this case is minimal.